openhands/runtime/browser/utils.py

```diff
@@ -103,13 +174,37 @@ async def browse(
             error=True if obs.get('last_action_error', '') else False,  # error flag
             trigger_by_action=action.action,
         )
+
+        # Process the content first using the axtree_object
+        observation.content = get_agent_obs_text(observation)
+
+        if not action.return_axtree:
+            # Then remove the axtree_object to save space
+            observation.dom_object = {}
+            observation.axtree_object = {}
+            observation.extra_element_properties = {}
+
+        return observation
     except Exception as e:
-        return BrowserOutputObservation(
-            content=str(e),
+        error_message = str(e)
+        error_url = asked_url if action.action == ActionType.BROWSE else ''
+
+        # Create error observation
+        observation = BrowserOutputObservation(
+            content=error_message,
             screenshot='',
             screenshot_path=None,
             error=True,
-            last_browser_action_error=str(e),
-            url=asked_url if action.action == ActionType.BROWSE else '',
+            last_browser_action_error=error_message,
+            url=error_url,
             trigger_by_action=action.action,
         )
+
+        if not action.return_axtree:
+            try:
+                observation.content = get_agent_obs_text(observation)
```

Comment:
This confuses me because it was already called on 179, and it relies on fields that were removed in 181. But I guess it's being called for the error message handling, so could make sense. Maybe if the part of `get_agent_obs_text` were extracted that does the error message and were called directly, but not a big deal. Also it's not clear to me why this is only done if return_axtree is False.

Reply:
> Also it's not clear to me why this is only done if return_axtree is False.

@OpenHands We should do `get_agent_obs_text` regardless of the value for `return_axtree`

Reply:
Should be ready for another round of review! @raymyers